x/lockup/ante.go

```diff
@@ -0,0 +1,132 @@
+func WrappedLockupAnteHandler(
+	anteHandler sdk.AnteHandler,
+	lockupKeeper keeper.Keeper,
+) sdk.AnteHandler {
+	wrapped := WrappedAnteHandler{anteHandler}
+	lad := NewLockupAnteDecorator(lockupKeeper)
+
+	return sdk.ChainAnteDecorators(wrapped, lad)
+}
+
+// LockupAnteDecorator Ensures that any transaction under a locked chain originates from a LockExempt address
+type LockupAnteDecorator struct {
+	lockupKeeper      keeper.Keeper
+	exemptSet         map[string]struct{}
+	lockedMsgTypesSet map[string]struct{}
+}
+
+// AnteHandle Ensures that any transaction under a locked chain originates from a LockExempt address
+func (lad LockupAnteDecorator) AnteHandle(ctx sdk.Context, tx sdk.Tx, simulate bool, next sdk.AnteHandler) (newCtx sdk.Context, err error) {
+	if lad.lockupKeeper.ChainIsLocked(ctx) {
+		for _, msg := range tx.GetMsgs() {
+			if lad.lockedMsgTypesSet == nil {
+				lad.initLockedMsgTypesSet(ctx)
+			}
+			if _, typePresent := lad.lockedMsgTypesSet[msg.Type()]; typePresent {
+				if lad.exemptSet == nil {
+					lad.initExemptSet(ctx)
+				}
+				if allow, err := allowMessage(msg, lad.exemptSet); !allow {
+					return ctx, sdkerrors.Wrap(err,
+						fmt.Sprintf("Transaction %v blocked because of message %v", tx, msg))
+				}
+			}
+		}
+	}
+	return next(ctx, tx, simulate)
+}
+
+// initLockedMsgTypesSet initializes the lockedMsgTypesSet
+func (lad LockupAnteDecorator) initLockedMsgTypesSet(ctx sdk.Context) {
+	lockedMsgTypes := lad.lockupKeeper.GetLockedMessageTypes(ctx)
+	lad.lockedMsgTypesSet = createSet(lockedMsgTypes)
+}
+
+func (lad LockupAnteDecorator) initExemptSet(ctx sdk.Context) {
+	exemptAddresses := lad.lockupKeeper.GetLockExemptAddresses(ctx)
+	lad.exemptSet = createSet(exemptAddresses)
+}
+
+// NewAnteHandler returns an AnteHandler that ensures any transaction under a locked chain
+// originates from a LockExempt address
+func NewLockupAnteHandler(lockupKeeper keeper.Keeper) sdk.AnteHandler {
+	return sdk.ChainAnteDecorators(NewLockupAnteDecorator(lockupKeeper))
+}
+
+// NewLockupAnteDecorator initializes a LockupAnteDecorator for locking messages
+// based on the settings stored in lockupKeeper
+// Note: Cannot init exemptSet nor lockedMsgTypesSet as the context may not be available yet
+func NewLockupAnteDecorator(lockupKeeper keeper.Keeper) LockupAnteDecorator {
+	return LockupAnteDecorator{lockupKeeper, nil, nil}
+}
+
+func createSet(strings []string) map[string]struct{} {
+	type void struct{}
+	var member void
+	set := make(map[string]struct{})
+
+	for _, str := range strings {
+		if _, present := set[str]; present {
+			continue
+		}
+		set[str] = member
+	}
+
+	return set
+}
+
+// allowMessage checks that an input `msg` was sent by only addresses in `exemptSet`
+// returns true if `msg` is either permissible or not a type of message this module blocks
+func allowMessage(msg sdk.Msg, exemptSet map[string]struct{}) (bool, error) {
+	switch msg.Type() {
+	case banktypes.TypeMsgSend:
```

Comment:
Should I just hardcode all the message types we're going to handle in here? Currently have a check that the message's type is in the lockedMsgTypesSet above in the AnteHandle method, then another check once we're in this method for the actual logic to evaluate the message. If we have a type in the lockedMsgTypesSet but no code handling it here, nothing gets blocked and I don't really like the duplication of concerns here.

Reply:
looking at that code it seems that the separation of concerns is only to ensure that not all messages make it to this function. If that's the case lets remove it, we can deal with a moderate efficiency solution and we'll work on optimizing if it we need to.

Reply:
We actually need the locked message types collection since it allows us to enable/disable checking per message type. But if a message's type does not have a case then we need to have a default behavior